File: frontend/src/services/productService.js
import api from './api';

export const productService = {
  async getAllProducts() {
    try {
      const response = await api.get('/api/products');
      return response.data;
    } catch (error) {
      throw error.response?.data || 'Failed to fetch products';
    }
  },

  async getProductById(id) {
    try {
      const response = await api.get(`/api/products/${id}`);
      return response.data;
    } catch (error) {
      throw error.response?.data || 'Failed to fetch product';
    }
  },

  async createProduct(productData) {
    try {
      const response = await api.post('/api/products', productData);
      return response.data;
    } catch (error) {
      throw error.response?.data || 'Failed to create product';
    }
  },

  async updateProduct(id, productData) {
    try {
      const response = await api.put(`/api/products/${id}`, productData);
      return response.data;
    } catch (error) {
      throw error.response?.data || 'Failed to update product';
    }
  },

  async deleteProduct(id) {
    try {
      const response = await api.delete(`/api/products/${id}`);
      return response.data;
    } catch (error) {
      throw error.response?.data || 'Failed to delete product';
    }
  },

  async getProductsByCategory(category) {
    try {
      const response = await api.get(`/api/products/category/${category}`);
      return response.data;
    } catch (error) {
      throw error.response?.data || 'Failed to fetch products by category';
    }
  },

  async getRecommendations(id) {
    try {
      const response = await api.get(`/api/products/${id}/recommendations`);
      return response.data;
    } catch (error) {
      throw error.response?.data || 'Failed to fetch recommendations';
    }
  }
};

File: backend/src/main/java/com/example/products/service/ProductService.java
package com.example.products.service;

import com.example.products.entity.Product;
import com.example.products.repository.ProductRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class ProductService {
    
    @Autowired
    private ProductRepository productRepository;

    public List<Product> getAllProducts() {
        return productRepository.findAll();
    }

    public Optional<Product> getProductById(Long id) {
        return productRepository.findById(id);
    }

    public List<Product> getProductsByCategory(String category) {
        return productRepository.findByCategory(category);
    }

    public List<Product> getRecommendations(Long productId) {
        Product product = productRepository.findById(productId)
                .orElseThrow(() -> new RuntimeException("Product not found"));
        
        return productRepository.findByCategoryAndIdNot(product.getCategory(), productId);
    }

    public Product saveProduct(Product product) {
        return productRepository.save(product);
    }

    public Product updateProduct(Long id, Product updatedProduct) {
        return productRepository.findById(id)
                .map(product -> {
                    product.setName(updatedProduct.getName());
                    product.setDescription(updatedProduct.getDescription());
                    product.setPrice(updatedProduct.getPrice());
                    product.setCategory(updatedProduct.getCategory());
                    product.setImageURL(updatedProduct.getImageURL());
                    return productRepository.save(product);
                })
                .orElseThrow(() -> new RuntimeException("Product not found"));
    }

    public void deleteProduct(Long id) {
        if (productRepository.existsById(id)) {
            productRepository.deleteById(id);
        } else {
            throw new RuntimeException("Product not found");
        }
    }
}
